Add replace-node-first-boot option

bhalevy opened the pull request:
Allow replacing a node given its Host ID rather than its ip address.

This series adds a replace_node_first_boot option to db/config
and makes use of it in storage_service.

The new option takes priority over the legacy replace_address* options.
When the latter are used, a deprecation warning is printed.

Documentation updated respectively.

And a cql unit_test is added.

Ref #12277

kbr-scylla commented on test/topology/test_topology.py:
```diff
@@ -99,15 +99,31 @@ async def no_joining_nodes():
+async def test_replace_using_host_id_reuse_ip(manager: ManagerClient, random_tables) -> None:
+    servers = await manager.running_servers()
+    await manager.server_stop(servers[0].server_id)
+    replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = True, use_host_id = True)
```
Maybe we could do the two replaces (using host ID and not using host ID) in a single test, so we don't grow the number of tests cases? Each test case requires starting a separate cluster - which takes some time.

(Maybe we should do the same with `reuse_ip_addr` - but out of this PR's scope)

bhalevy replied:
okay.

Maybe we should take the same approach in
https://github.com/scylladb/scylla-dtest/pull/2946

@fruch what do you think?

bhalevy replied:
@kbr-scylla this has proven to be more difficult than I thought, since it the second replace in a row fails (ta least some of the times) because it sees one of the nodes as down.  I think it's the first replacing node that is not recognized as UN by all other nodes when add_server resolves.

It's out of scope for this PR to deal with that, so I'll keep the tests separated and we can merge them as a follow up.

kbr-scylla replied:
Ok. Maybe I'll work on merging them later.

kbr-scylla commented on docs/operating-scylla/procedures/cluster-management/replace-dead-node.rst:
```diff
@@ -189,11 +189,11 @@ In this case, the node's data will be cleaned after restart. To remedy this, you
       sudo sed -e '/auto_bootstrap:.*/s/False/True/g' -i /etc/scylla/scylla.yaml
 
-#. Run the following command, replacing 172.30.0.186 with the listen_address / rpc_address of the node that you are restarting:
+#. Run the following command to replace the Host ID of the node that you are restarting:
```
The previous statement instructed the reader to take the command below, modify the command by replacing the IP provided here with the IP of the node that they want to restart, and then run the command.

This statement tells the reader that the command replaces the Host ID of the node that the reader is restarting... replaces it with what? This doesn't make much sense. And running the command verbatim is obviously an error - we don't want the reader to run the following command, actually (we want them to run a modified command).

I think we should keep the format of the statement as it was and write something like:
"Run the following command, replacing 87a2b646-c968-4108-a38b-f8468b8b7360 with the Host ID of the node that you are restarting"

bhalevy replied:
yeah, that makes sense.

bhalevy replied:
I'm actually not happy with the phrase you suggested since the action is not about replacing the Host ID, it's about replacing the node previously known by that Host ID.

How about:
```
#. Run the following command to replace the node known by the previous Host ID of the node you are restarting, 87a2b646-c968-4108-a38b-f8468b8b7360, by the restarted instance, that will be assigned a new, random, Host ID:
```

kbr-scylla replied:
> action is not about replacing the Host ID

The command itself is not about replacing the Host ID, but the reader of the documentation must first replace the Host ID in the command with the Host ID of their node. That's what the "replacing 87a2b646-c968-4108-a38b-f8468b8b7360 with..." is about.

So the statement is a succint wait of saying:
1. In the command below, replace "87a2b646-c968-4108-a38b-f8468b8b7360" with the Host ID of the node that you are restarting
2. Then execute the command below.

kbr-scylla commented on init.cc:
```diff
@@ -50,6 +50,10 @@ std::set<gms::inet_address> get_seeds_from_db_config(const db::config& cfg) {
+        startlog.error("Bad configuration: replace-node-first-boot is not allowed for seed nodes");
+        throw bad_configuration_error();
+    }
     if ((!cfg.replace_address_first_boot().empty() || !cfg.replace_address().empty()) && seeds.contains(broadcast_address)) {
```
Why not include `!cfg.replace_node_first_boot().empty()` in this alternative (and modify the error contents)? It would be more consistent to either put all the options together since the existing options were already together.

bhalevy replied:
It's possible, but it will mean less churn once the support for the old options is dropped.
Also the inner logic will have to check `cfg.replace_node_first_boot().empty()` again to control the error message.

kbr-scylla commented on service/storage_service.hh:
```diff
@@ -295,9 +295,11 @@ private:
+    future<> prepare_replacement_info(std::unordered_set<gms::inet_address> initial_contact_nodes,
             const std::unordered_map<gms::inet_address, sstring>& loaded_peer_features);
 
+    std::optional<replacement_info> _replacement_info;
```
I won't object to this, but I don't like putting more state inside the class :( It's not clear if this state is mutable, where it's needed etc. Would be nice if we could avoid it by just passing the `replacement_info` where we need it.

bhalevy replied:
I agree in principle, but unfortunately, the join_token_ring, handle_state_replacing song and dance requires this state :-(

bhalevy replied:
not present in v5, instead
the local replacement_info prepared in join_token_ring is used to pass the replace address down to functions that need it and are called from join_token_ring

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -171,19 +171,10 @@ void storage_service::register_metrics() {
-    } catch (...) {
-        return std::nullopt;
+    if (!_replacement_info) {
+        on_internal_error(slogger, "replacement_info is unset");
```
You could quickly check it in your test.py tests by (temporarily just to see, I don't suggest to send a patch with this) including `'enable_repair_based_node_ops': False` in `make_scylla_conf` in pylib/scylla_cluster.py

bhalevy replied:
Still, main -> join_cluster -> join_token_ring must complete before on_change -> handle_state_replacing happens.
The event-based state machine relies on gossip to transition the replacing node between those states.
Other nodes will process the transition of the replacing node as part of on_change -> join_cluster.

Is that right @asias?

FWIW, test_topology passes with `'enable_repair_based_node_ops': False in make_scylla_conf`

asias replied:
with/without enable_repair_based_node_ops == true  for replace, handle_state_replacing is not going to be called anymore, since the replacing node will not publish the HIBERNATE state any more.

asias replied:
A node with new code running replace will not publish HIBERNATE state, so handle_state_replacing on new and old node will not be called. 

But if an old node runs replacing, it will publish HIBERNATE, so handle_state_replacing will be called. In this case, the replacing node will use the same host_id as the node to be replaced anyway. So even if in your make host id unique patch, it does not handle the case replacing node uses new host id in handle_state_replacing , it should be fine.

kbr-scylla replied:
> Still, main -> join_cluster -> join_token_ring must complete before on_change -> handle_state_replacing happens.

handle_state_replacing will also happen on other nodes, not only the one which is replacing. These nodes will not have `_replacement_info`.

> But if an old node runs replacing, it will publish HIBERNATE, so handle_state_replacing will be called.

Let's assume that we don't run replace in the middle of upgrade, so in particular we don't run replace with "old nodes".

But that doesn't change anything, the user can still use `'enable_repair_based_node_ops': False` in a **new node**, and that node will publish HIBERNATE when replacing, which will cause other nodes to call `handle_state_replacing`, right?

kbr-scylla replied:
If `'enable_repair_based_node_ops': False` is a legal thing to do, we must continue supporting it. Only once we completely get rid of the option, we can assume that nodes won't publish HIBERNATE and so on.

bhalevy replied:
I don't understand how does the current code supposed work with get_replace_address.
It would return a disengaged optional on nodes other than the replacing one.
We can support that

kbr-scylla replied:
https://github.com/scylladb/scylladb/issues/12330

kbr-scylla replied:
> It would return a disengaged optional on nodes other than the replacing one.

Yes.

But after understanding that `handle_state_replacing` should no longer be ever called, I'm proposing to remove it (outside of this PR), and then we'll easily see that `get_replace_address` is called only by the replacing node.

For now, I'm okay with the runtime check - `handle_state_replacing` can be broken since it's dead code anyway.

bhalevy replied:
works for me

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3501/

bhalevy commented on service/storage_service.cc:
```diff
@@ -187,12 +187,24 @@ std::optional<gms::inet_address> storage_service::get_replace_address() {
+        return true;
+    }
+
+    if (!cfg.replace_address_first_boot().empty() || !cfg.replace_address().empty()) {
```
Same as in init.cc:
1) easier path to removing support for the legacy config options
2) simpler logic overall IMO since to control the error message we'll need to check again which of the cfg options is given.

asias replied:
This way when we drop the replace_address_first_boot option, we can drop the old code without touching the new code.  It is also either to reason about the code since less conditions in a single if.

bhalevy replied:
I will combine them in the next version, as I realized that `cfg.replace_address().empty()` should be handled separately to follow the existing the logic (and cause an exception down the road if `!cfg.replace_address().empty()` and `_sys_ks.local().bootstrap_complete()`)

bhalevy replied:
> This way when we drop the replace_address_first_boot option, we can drop the old code without touching the new code. It is also either to reason about the code since less conditions in a single if.

You managed to convince me to flip my mind again :)

bhalevy commented on service/storage_service.cc:
```diff
@@ -1642,14 +1642,24 @@ future<> storage_service::remove_endpoint(inet_address endpoint) {
+        replace_address = gms::inet_address(cfg.replace_address());
+        slogger.warn("The replace_address={} option is deprecated. Please use the replace_node_first_boot option", replace_address);
+    } else {
+        throw std::runtime_error(format("No replace node or replace address configuration options found"));
```
right, will change

bhalevy commented on service/storage_service.cc:
```diff
@@ -1680,11 +1701,17 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
-    auto replace_host_id = _gossiper.get_host_id(replace_address);
+    if (!replace_host_id) {
+        replace_host_id = _gossiper.get_host_id(replace_address);
+    } else if (replace_host_id != _gossiper.get_host_id(replace_address)) {
```
just a sanity check for gossip about the back and forth mapping, but we can do without it.

bhalevy commented on service/storage_service.cc:
```diff
@@ -1680,11 +1701,17 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
     co_await _gossiper.reset_endpoint_state_map();
 
     _replacement_info = replacement_info {
+        .replace_host_id = replace_host_id,
```
yeah, good catch.

asias commented on service/storage_service.hh:
```diff
@@ -291,13 +291,18 @@ private:
     future<> shutdown_protocol_servers();
 
     struct replacement_info {
+        locator::host_id replace_host_id;
```
replace_host_id is a duplication of  host_id below.

bhalevy replied:
fixed in the next version

asias commented on service/storage_service.hh:
```diff
@@ -291,13 +291,18 @@ private:
 
     struct replacement_info {
+        locator::host_id replace_host_id;
+        gms::inet_address replace_address;
```
This is inside replacement_info, just call it address like other members without the replace prefix?

bhalevy replied:
right, done this way in next version

asias commented:
I suggest adding a warning if the old replace options are used and point the user to use the new option instead.

bhalevy commented:
> I suggest adding a warning if the old replace options are used and point the user to use the new option instead.

I added these warnings in prepare_replacement_info:
```c++
    } else if (!cfg.replace_address_first_boot().empty()) {
        replace_address = gms::inet_address(cfg.replace_address_first_boot());
        slogger.warn("The replace_address_first_boot={} option is deprecated. Please use the replace_node_first_boot option", replace_address);
    } else if (!cfg.replace_address().empty()) {
        replace_address = gms::inet_address(cfg.replace_address());
        slogger.warn("The replace_address={} option is deprecated. Please use the replace_node_first_boot option", replace_address);
```

bhalevy commented:
In v2 ([83e64db](https://github.com/scylladb/scylladb/commit/83e64dbd7099b0bf0c82cdbdea953edd0458f08c)):
- rebased
- storage_service: prepare_replacement_info: keep the replaced node host_id and address in replacement_info
  - add only `.address` member
  - reuse .host_id
- storage_service: is_replacing: rely directly on config options
  - doesn't change logic for replace_address config option handling
    - will cause exception to be thrown if node is already bootstrapped
    - rather than returning false for is_replacing
- storage_service: get_replace_address: use _replacement_info
  - Improved git log
- test: pylib: ServerInfo: add host_id
  - new
  - isn't used at this point in test_topology
  - but can be used later for e.g. providing ignore_dead_nodes_for_replace
- db: config: describe replace_address* options as deprecated
  - fix typo in ignore_dead_nodes_for_replace description
- docs: document the new replace_node_first_boot option
  - Rephrased "Run the following command" as per https://github.com/scylladb/scylladb/pull/12316#discussion_r1049254197

annastuchlik commented on docs/operating-scylla/procedures/cluster-management/replace-dead-node.rst:
```diff
@@ -189,11 +189,11 @@ In this case, the node's data will be cleaned after restart. To remedy this, you
       sudo sed -e '/auto_bootstrap:.*/s/False/True/g' -i /etc/scylla/scylla.yaml
 
-#. Run the following command, replacing 172.30.0.186 with the listen_address / rpc_address of the node that you are restarting:
+#. Run the following command to replace the node known by the previous Host ID of the node you are restarting, 87a2b646-c968-4108-a38b-f8468b8b7360, by the restarted instance, that will be assigned a new, random, Host ID:
```
```suggestion
#. Run the following command to replace the dead node (previously known by the Host ID of the node you are restarting) with the restarted node:
```

annastuchlik replied:
I've split and rephrased the sentence because it was hard to process. Also, I think the Host ID is just an example, so we should remove it from the instructions.

bhalevy replied:
The thing is that it is not officially dead.
The scenario here is that this node lost its ephemeral storage, so referring to the old incarnation as "dead' could be misleading.
Maybe you can think of a clearer way to describe it.

annastuchlik replied:
I see your point. How about:

Run the following command to replace the instance whose ephemeral volumes were erased (previously known by the Host ID of the node you are restarting) with the restarted instance. The restarted node will be assigned a new random Host ID.

kbr-scylla replied:
TBH I don't see a point in explaining what the command does here. The previous version did not explain it either.

And my head hurts when I read these complex sentences.

How about:

"In the following command, replace `87a2b646-c968-4108-a38b-f8468b8b7360` with the Host ID of the node that you are restarting. Then run the command:"

Optionally we can explain what the command is needed for, but not in this paragraph. Explain it in a separate paragraph before this one or after the code block.

bhalevy replied:
Ah, now I understand what you initially meant.

As far as I can tell, in the documented instructions we currently provided, we don't instruct the readers what they should do to actually execute our instructions, but rather we provide examples and explain what the examples do, taking it for granted that admins will replace the "arguments" used in the example commands with the right ones in their environment.

You instructions were understood by me as explaining to the admin that the command is used to replace the node's Host ID with some other Host ID, which doesn't make sense.

The alternative, maybe as a followup cleanup, is to replace the actual bogus arguments like particular ip addresses or (sometimes invalid, sometimes valid) host UUIDs with abstract arguments like `<node-X-ip-address>` and `<node-X-host-id>`

bhalevy replied:
that way it's clear that these are placeholders for actual arguments

annastuchlik replied:
I'd suggest leaving the section as is to avoid blocking this PR and refactor it later, but it's up to you. My thoughts on refactoring:

- Complex sentences are a pain, so I split the instructions in that step to make it read better. However, I wouldn't give up explaining to the user WHY they need to run a given command. It doesn't have to be explained in that very step - there could be a separate paragraph preceding the instructions, but we shouldn't leave the users in the dark about what a given command does.
- About the examples: I don't think the current example is confusing, but there's a better way. When I add an example, I tend to add a command with a placeholder followed by an example. For example:

Take a snapshot, including every keyspace you want to backup:

`nodetool snapshot <KEYSPACE_NAME>`

For example: 
	
`nodetool snapshot mykeyspace`

bhalevy replied:
that's a good approach

annastuchlik commented on docs/operating-scylla/procedures/cluster-management/replace-dead-node.rst:
```diff
@@ -68,11 +68,11 @@ Procedure
     For example (using the address of the failed node from above):
 
-    ``replace_address_first_boot: 192.168.1.203``   
+    ``replace_node_first_boot: 87a2b646-c968-4108-a38b-f8468b8b7360``
```
@bhalevy 
- I understand that Host ID should be used, but I'm not sure why this one is. The Host ID of the dead node in the example (address 192.168.1.203) is 675ed9f4-6564-6dbd-can8-43fddce952gy.
-I think that "address" on line 73 should be replaced with "Host ID".

bhalevy replied:
> @bhalevy
> 
> * I understand that Host ID should be used, but I'm not sure why this one is. The Host ID of the dead node in the example (address 192.168.1.203) is 675ed9f4-6564-6dbd-can8-43fddce952gy.

good point

> * I think that "address" on line 73 should be replaced with "Host ID".

Right, thanks

bhalevy commented:
In v3 ([ea57b93](https://github.com/scylladb/scylladb/commit/ea57b932c5516ff54a2d332638b8ffc6a06ec498)):
- addressed docs review comments

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3520/

annastuchlik reviewed (approved):
The docs look good.

kostja commented on docs/operating-scylla/admin.rst:
```diff
@@ -80,8 +80,8 @@ The following addresses can be configured in scylla.yaml:
-   * - replace_address_first_boot
-     - Address of the node this Scylla instance is meant to replace. Refer to :doc:`Replace a Dead Node in a Scylla Cluster </operating-scylla/procedures/cluster-management/replace-dead-node>` for more details.
+   * - replace_node_first_boot
+     - Host ID of the node this Scylla instance is meant to replace. Refer to :doc:`Replace a Dead Node in a Scylla Cluster </operating-scylla/procedures/cluster-management/replace-dead-node>` for more details.
```
Just in this sentence we use node, instance and host for the same thing. Could we settle on fewer names for the same thing perhaps or provide a glossary of what the differences are? At the very least we perhaps should consistently use "Scylla node" rather than "instance" in all places.

bhalevy replied:
Makes sense.  How about:

```
  * - replace_node_first_boot
     - Host ID of a dead node this Scylla node is replacing. Refer to ...
```

bhalevy replied:
Rephrased in [6808aa1](https://github.com/scylladb/scylladb/commit/6808aa1bda5f9f464683dd1e4e5a158d17221344)

kbr-scylla replied:
It still says "Host ID of the node this Scylla instance is meant to replace"

kostja commented on docs/operating-scylla/procedures/cluster-management/replace-dead-node.rst:
```diff
@@ -189,11 +189,11 @@ In this case, the node's data will be cleaned after restart. To remedy this, you
       sudo sed -e '/auto_bootstrap:.*/s/False/True/g' -i /etc/scylla/scylla.yaml
 
-#. Run the following command, replacing 172.30.0.186 with the listen_address / rpc_address of the node that you are restarting:
+#. Run the following command to replace the instance whose ephemeral volumes were erased (previously known by the Host ID of the node you are restarting) with the restarted instance. The restarted node will be assigned a new random Host ID.
```
Same comment re instance vs host vs node.

bhalevy replied:
Here instance refers to a cloud VM that lost its ephemeral storage, so I think it does make sense in this context, but it can probably be said more clearly.  How about:

```
To restore the data lost by the restarted instance, locate the Host ID of the respective Scylla node and use the following option when restarting the node to initiate its replacement. The restarted node will be assigned a new random Host ID.
```

kostja replied:
Maybe I am mistaken, but I think we're not restoring any data when replacing? We're restoring the replication factor, but the data is streamed from other nodes, it's not taken from some volume locally? Now I don't get how can you *restart* a node yet change its startup options, given its instance (VM) was lost? Rather, we're starting a fresh new node with intention to replace the node lost along with the permanently lost VM. Am I getting this right?

bhalevy replied:
It comes down to the terminology, that I don't think our documentation is crispy clear about.
This scenario calls for a cloud instance that loses the data volume.  When restarted, scylla will pick a new random host_id and the replace option (to scylla, not to the instance) tells it to replace the dead node that was lost, that happened to be served on the same instance, but it's almost irrelevant (except maybe for reusing the same ip address).
The node's content is rebuilt from the remaining replicas.

kostja replied:
In that case I think it would be more clear to say that we're replacing the node and stream its data set from the remaining replicas. The only thing that stays the same is IP address and tokens.

bhalevy replied:
Rephrased in [07d8629](https://github.com/scylladb/scylladb/commit/07d8629e2bd922483cb2000d26f5bf285def0b9a)

kostja commented on db/config.cc:
```diff
@@ -788,6 +788,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , consistent_rangemovement(this, "consistent_rangemovement", value_status::Used, true, "When set to true, range movements will be consistent. It means: 1) it will refuse to bootstrap a new node if other bootstrapping/leaving/moving nodes detected. 2) data will be streamed to a new node only from the node which is no longer responsible for the token range. Same as -Dcassandra.consistent.rangem ...
     , join_ring(this, "join_ring", value_status::Unused, true, "When set to true, a node will join the token ring. When set to false, a node will not join the token ring. User can use nodetool join to initiate ring joinging later. Same as -Dcassandra.join_ring in cassandra.")
     , load_ring_state(this, "load_ring_state", value_status::Used, true, "When set to true, load tokens and host_ids previously saved. Same as -Dcassandra.load_ring_state in cassandra.")
+    , replace_node_first_boot(this, "replace_node_first_boot", value_status::Used, "", "The Host ID of the dead node to replace. If this node has been bootstrapped successfully, this option will be ignored.")
```
What is this node? Is "this node" the node to replace or the node that has been started with this option? Would be really nice to disambiguate.

bhalevy replied:
It's the replacing node. How about the following:

The Host ID of a dead node to replace. If the replacing node has already been bootstrapped successfully, this option will be ignored.

bhalevy replied:
Rephrased in [07d8629](https://github.com/scylladb/scylladb/commit/07d8629e2bd922483cb2000d26f5bf285def0b9a)

kbr-scylla replied:
It still says "this node".

bhalevy replied:
oops, will change.

bhalevy commented:
Converted to draft until #12330 is merged.

bhalevy commented:
In v5 ([dbfc33d](https://github.com/scylladb/scylladb/commit/dbfc33d9fce5a90bdce371cd68c1182344202a6d)):
- rebased on top of 767b7be8be1b736c8e098684e7b6870661b1b26b
- with no need for a storage_service::_replacement_info member
- revised and new patches:
  - storage_service: replacement_info: add replace address
  - storage_service: join_token_ring: reuse replacement_info.address
  - storage_service: pass replacement_info to booststrap
  - storage_service: pass replace_address to run_replace_ops
  - storage_service: get rid of get_replace_address

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -499,12 +499,10 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
         } else {
-            auto replace_addr = get_replace_address();
-            assert(replace_addr);
-            if (*replace_addr != get_broadcast_address()) {
```
Maybe `assert(replace_address)`?

bhalevy replied:
Or change `if (is_replacing())` to `if (replace_address)` since it's set iff the former is true.

kbr-scylla replied:
Sounds good.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -732,15 +733,15 @@ future<> storage_service::bootstrap(cdc::generation_service& cdc_gen_service, st
         } else {
-            auto replace_addr = get_replace_address();
-            assert(replace_addr);
+            auto replace_addr = replacement_info->address;
```
Maybe `assert(replacement_info)`?

bhalevy replied:
can also check `if (replacement_info)` rather than `if (is_replacing())`

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -187,12 +187,27 @@ std::optional<gms::inet_address> storage_service::get_replace_address() {
 }
 
 bool storage_service::is_replacing() {
```
Re commit message (`storage_service: is_replacing: rely directly on config options
`): "before we change the latter..." -> "before we remove the latter".

bhalevy replied:
yup, will edit it.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -187,12 +187,27 @@ std::optional<gms::inet_address> storage_service::get_replace_address() {
+    }
+    // Returning true if cfg.replace_address is provided
+    // will trigger an exception down the road if bootstrap_complete(),
+    // as it is an error to use this option post bootstrap.
```
I understand this is a pre-existing problem, but maybe we should just throw an exception right here. (optional)

BTW. the documentation of "replace_address" is misleading when compared to "replace_address_first_boot", it suggests that we might be able to replace using a node that has already bootstrapped... Did this ever actually work? Did it actually work in Cassandra? It doesn't make any sense to me.

bhalevy replied:
I'm not sure how it worked.
It's discouraged and I'm not sure why we even bother to support it anymore.

kbr-scylla reviewed (approved):
LGTM, minor comments.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3678/

bhalevy commented:
In v6 ([91e696f](https://github.com/scylladb/scylladb/commit/91e696fc0819d49c4b839a78cbf3936e8f45c946)):
- db: config: add replace_node_first_boot option
  - fixed description
- storage_service: join_token_ring: reuse replacement_info.address
  - check for `bool(replace_address)` rather than `is_replacing()`
- storage_service: pass replacement_info to booststrap
  - check for `bool(replacement_info)` rather than `is_replacing()`
- storage_service: is_replacing: rely directly on config options
  - fixed commit message

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3687/

kbr-scylla commented:
`boost.mutation_reader_test.combined_reader_galloping_within_partition_test` failed, apparently there was a segfault? Nothing interesting in the logs except exit code:
https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3687/artifact/testlog/x86_64/dev/boost.mutation_reader_test.combined_reader_galloping_within_partition_test.1.log

cc @denesb 

Restarting the CI

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3696/

denesb commented:
> `boost.mutation_reader_test.combined_reader_galloping_within_partition_test` failed, apparently there was a segfault? Nothing interesting in the logs except exit code: https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3687/artifact/testlog/x86_64/dev/boost.mutation_reader_test.combined_reader_galloping_within_partition_test.1.log
> 
> cc @denesb
> 
> Restarting the CI

Looks like https://github.com/scylladb/scylladb/issues/11842.

bhalevy commented:
@kbr-scylla this series got dequeued due to topology tests flakiness.
Can we try again?
It blocks https://github.com/scylladb/scylladb/pull/12506

kbr-scylla commented:
Yes, the problem was too many test causing the test framework to run out of IPs, but that was fixed already.
Could you please rebase?

bhalevy commented:
v7 ([de3142e](https://github.com/scylladb/scylladb/commit/de3142e5402f85c100f7f7187b41d4a4b49b00de)):
- rebased
- pass replacement_info to run_replace_ops rather than just the endpoint
  - anticipating https://github.com/scylladb/scylladb/pull/12506

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3838/
